tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp

```diff
@@ -0,0 +1,310 @@
+    auto it = std::find_if(
+        apis.begin(), apis.end(),
+        [&](const model::Api& obj) -> bool { return obj.GetApi() == service; });
+
+    write::model::Apis result;
+    if (it != apis.end()) {
+      result.push_back(*it);
+    }
+
+    return result;
+  }
+
+  std::shared_ptr<NetworkMock> network_;
+  olp::client::OlpClientSettings settings_;
+};
+
+TEST_F(VersionedLayerClientTest, StartBatch) {
+  const auto catalog = kHrn.ToCatalogHRNString();
+  const auto api = CreateApiResponse(kPublishApiName);
+  const auto publication =
+      mockserver::DefaultResponses::GeneratePublicationResponse({kLayer}, {});
+  ASSERT_FALSE(api.empty());
+
+  {
+    SCOPED_TRACE("Success request");
+
+    EXPECT_CALL(*network_, Send(IsGetRequest(kLookupPublishApiUrl), _, _, _, _))
+        .WillOnce(ReturnHttpResponse(olp::http::NetworkResponse().WithStatus(
+                                         olp::http::HttpStatusCode::OK),
+                                     olp::serializer::serialize(api)));
+    EXPECT_CALL(*network_, Send(IsPostRequest(kPublishUrl), _, _, _, _))
+        .WillOnce(ReturnHttpResponse(olp::http::NetworkResponse().WithStatus(
+                                         olp::http::HttpStatusCode::OK),
+                                     olp::serializer::serialize(publication)));
+
+    write::VersionedLayerClient write_client(kHrn, settings_);
+    const auto batch_request = model::StartBatchRequest().WithLayers({kLayer});
+    const auto response =
+        write_client.StartBatch(batch_request).GetFuture().get();
+    const auto& result = response.GetResult();
+
+    EXPECT_TRUE(response.IsSuccessful());
+    ASSERT_TRUE(result.GetId());
+    ASSERT_TRUE(result.GetDetails());
+    ASSERT_TRUE(result.GetLayerIds());
+    ASSERT_EQ(result.GetLayerIds().get().size(), 1);
+    ASSERT_EQ(result.GetLayerIds().get().front(), kLayer);
+    ASSERT_NE("", response.GetResult().GetId().value());
+    Mock::VerifyAndClearExpectations(network_.get());
+  }
+
+  {
+    SCOPED_TRACE("Success request, but with callback");
+
+    // don't mock lookup api request since its cached in previous step
+    EXPECT_CALL(*network_, Send(IsPostRequest(kPublishUrl), _, _, _, _))
+        .WillOnce(ReturnHttpResponse(olp::http::NetworkResponse().WithStatus(
+                                         olp::http::HttpStatusCode::OK),
+                                     olp::serializer::serialize(publication)));
+
+    std::promise<write::StartBatchResponse> promise;
+    auto future = promise.get_future();
+
+    write::VersionedLayerClient write_client(kHrn, settings_);
+    const auto batch_request = model::StartBatchRequest().WithLayers({kLayer});
+    const auto token = write_client.StartBatch(
+        batch_request, [&promise](write::StartBatchResponse response) {
+          promise.set_value(std::move(response));
+        });
+    const auto response = future.get();
+    const auto& result = response.GetResult();
```

Comment:
Same here please separate the request and the waiting:

```suggestion
    std::promise<write::StartBatchResponse> promise;
    write::VersionedLayerClient write_client(kHrn, settings_);
    const auto batch_request = model::StartBatchRequest().WithLayers({kLayer});
    const auto token = write_client.StartBatch(
        batch_request, [&promise](write::StartBatchResponse response) {
          promise.set_value(std::move(response));
        });
        
     auto future = promise.get_future();   
    const auto response = future.get();
    const auto& result = response.GetResult();
```